File: BoardGameGeek/src/com/boardgamegeek/provider/ArtistsProvider.java
package com.boardgamegeek.provider;

import android.net.Uri;

import com.boardgamegeek.provider.BggContract.Artists;
import com.boardgamegeek.provider.BggDatabase.Tables;

public class ArtistsProvider extends BasicProvider {

	@Override
	protected String getDefaultSortOrder() {
		return Artists.DEFAULT_SORT;
	}

	@Override
	protected String getInsertedIdColumn() {
		return Artists.ARTIST_ID;
	}

	@Override
	protected String getPath() {
		return BggContract.PATH_ARTISTS;
	}

	@Override
	protected String getTable() {
		return Tables.ARTISTS;
	}

	@Override
	protected String getType(Uri uri) {
		return Artists.CONTENT_TYPE;
	}
}

File: app/src/main/java/com/boardgamegeek/provider/ArtistsProvider.kt
package com.boardgamegeek.provider

import android.net.Uri
import android.provider.BaseColumns

import com.boardgamegeek.provider.BggContract.Artists
import com.boardgamegeek.provider.BggContract.Companion.PATH_ARTISTS
import com.boardgamegeek.provider.BggDatabase.Tables

class ArtistsProvider : BasicProvider() {
    override fun getType(uri: Uri) = Artists.CONTENT_TYPE

    override val path = PATH_ARTISTS

    override val table = Tables.ARTISTS

    override val defaultSortOrder = Artists.DEFAULT_SORT

    override val insertedIdColumn = Artists.Columns.ARTIST_ID

    override fun buildExpandedSelection(uri: Uri, projection: Array<String>?): SelectionBuilder {
        val builder = SelectionBuilder()
            .mapToTable(BaseColumns._ID, table)
            .mapToTable(Artists.Columns.ARTIST_ID, table)
            .mapToTable(Artists.Columns.UPDATED, table)
        if (projection.orEmpty().contains(Artists.Columns.ITEM_COUNT)) {
            builder
                .table(Tables.ARTISTS_JOIN_COLLECTION)
                .groupBy("$table.${Artists.Columns.ARTIST_ID}")
                .mapAsCount(Artists.Columns.ITEM_COUNT)
        } else {
            builder.table(table)
        }
        return builder
    }
}
